### DISCUSSION TITLE
Dynamic error messages in model validations do not work correctly

### DISCUSSION THREAD

**Discussion by tapirus** [wheels-dev/wheels#1155] [Q&A]
Bug description:
My app is localized and uses multiple languages. 
The localized strings are used like this. The session.locale may be "en_US", "fr_FR" etc.
#$('global.This is a text.',session.locale)#

This works perfectly, the only exception are the model error messages.
The session.locale is ignored and a random language string is shown.

The corresponding code in wheels is mentioned here:
https://github.com/cfwheels/cfwheels/issues/470
and 
https://github.com/cfwheels/cfwheels/commit/eb331981bd61475ddd65471beff33d624083dd85

I guess this line of code is not correct in wheels/model/validations.cfm
loc.rv = Evaluate(DE(loc.rv));

Why there is the "DE" included?
Maybe a short example or description how to use localized language strings would clarify?

Use custom model validation like this:
validatesFormatOf(property="SystemID", condition="this.systemid EQ ''", 
			message="#$('actors.Please select the system.',session.locale)#");

If you have more than one localized language. The wheels will pick a language randomly and not of the actual session.locale.

**Expected behavior**
If you use a custom model validation like this, the error message should be shown in the correct language.
validatesFormatOf(property="SystemID", condition="this.systemid EQ ''", 
			message="#$('actors.Please select the system.',session.locale)#");

**Additional context**
This refers to all the latest wheels versions.

A fix would be nice.

**Comment by chapmandu** [comment]
The models are cached, so I have a feeling that whoever's session made the request that called the onapplicationstart event, would be the language that is used.

You may have to think about using a global function or service to do the translation and move the session reference out of the model.

**Comment by neokoenig** [comment]
The config is cached, but you can just use a custom validator to call the function from outside the config();
```
config(){
  validate("myCustomValidation");
}

function myCustomValidation(){
 if(structKeyExists(this, "foo") && this.foo == 'badthing'){ 
   this.addError(property="foo", message=localise('Foo Error Msg'));
  }
}
```

`localise` in this context would be a wrapper/helper function in the global/functions.cfm file which would then call your localisation service etc.

**Comment by bpamiri** [comment]
I converted this into a discussion. @tapirus see if the answers from @neokoenig or @chapmandu solve your issue and mark the question answered. If we still need to turn this into an issue to address we can do so.

**Reply by tapirus** [reply]
Hi Adam, Tom and Peter,
thanks for helping out.
1. I guess Toms recommended workaround will work. Thanks.
2. But if this does mean I have to convert each and every standard wheels validation to a custom validation just to have a localized error message, it does not seem very wheels like for me. It will bloat the code a lot. Does it not make sense to implement a simple solution to the wheels core to make it easier for us?
What do you think? I don't know how much work would this be.
Thanks
Thorsten

**Reply by neokoenig** [reply]
I should clarify, you don't have use a custom validation; but you do need to user a wrapper function around the message text to your localisation function to prevent anything `session.` related in the `config()` block.

**Reply by tapirus** [reply]
Thanks for clarification. I got it working with custom addError localised messages, but still not with standard wheels validations.

In the model I inserted in config()
validate("customValidations");

in global/functions.cfm I have
public function localise(required string textstring) {
        return($('#textstring#',session.locale));
}

in the model are the following variants of validations for the same property:
// Custom validations outside config() to not be cached
	private void function customValidations() {
     // is not recognized - instead default wheels error message is shown
    validatesInclusionOf(property="GenderID", list="1,2", 
			message=localise('therapists.Please select the title.'));

	// does work
	 if(structKeyExists(this, "GenderID") && this.GenderID == ''){ 
   		this.addError(property="GenderID", message=localise('therapists.Please select the title.'));
   	}
}
Why does the custom addError function work and the standard validatsInclusionOf() not? 
What am I missing?

**Reply by tapirus** [reply]
Hi,
I can't solve the above mentioned problem.

Why are standard wheels validations outside of config() in the model not recognized by wheels?
In my model I call 
validate("customValidatesPresenceOfFirstName");

Outside of config() I have this:
function customValidatesPresenceOfFirstName(){
	validatesPresenceOf(property="FirstName", message=localise('therapists.Please enter the first name.'));
}

But unfortenately the validation is not recognized by wheels.
Instead I get the default wheels validation on this property.

Does anybody know why?

**Reply by chapmandu** [reply]
You cannot use `validatesPresenceOf` outside of `config()`.

Your custom validator is just that.. you need to do something like:

```
function customValidatesPresenceOfFirstName(){
  if (this.propertyIsBlank("FirstName")) {
    this.addError(property="FirstName", message=localise('therapists.Please select the first name.'));
  }
}
```
